presto-main/src/main/java/io/prestosql/operator/scalar/ConcatWsFunction.java

```diff
@@ -0,0 +1,211 @@
+            if (!(arrayValueType instanceof VarcharType) && !(arrayValueType instanceof UnknownType)) {
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Array values must be varchar types");
+            }
+
+            return new ChoicesScalarFunctionImplementation(
+                    functionBinding,
+                    NULLABLE_RETURN,
+                    ImmutableList.of(NEVER_NULL, NEVER_NULL),
+                    methodHandle(ConcatWsFunction.class, "concatWsArray", Slice.class, Block.class));
+        }
+    }
+
+    public ConcatWsFunction()
+    {
+        super(new FunctionMetadata(
+                new Signature(
+                        "concat_ws",
+                        ImmutableList.of(),
+                        ImmutableList.of(),
+                        VARCHAR.getTypeSignature(),
+                        ImmutableList.of(VARCHAR.getTypeSignature(), VARCHAR.getTypeSignature()),
+                        true),
+                true,
+                ImmutableList.of(new FunctionArgumentDefinition(false), new FunctionArgumentDefinition(true)),
+                false,
+                true,
+                "Concatenates elements using separator",
+                SCALAR));
+    }
+
+    @Override
+    public ScalarFunctionImplementation specialize(FunctionBinding binding)
+    {
+        if (binding.getArity() < 2) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "There must be two or more arguments");
+        }
+
+        int valueCount = binding.getArity() - 1;
+        MethodHandle arrayMethodHandle = methodHandle(ConcatWsFunction.class, "concatWs", Slice.class, Slice[].class);
+        MethodHandle customMethodHandle = arrayMethodHandle.asCollector(Slice[].class, valueCount);
+
+        return new ChoicesScalarFunctionImplementation(
+                binding,
+                NULLABLE_RETURN,
+                ImmutableList.<InvocationConvention.InvocationArgumentConvention>builder()
+                        .add(NEVER_NULL)
+                        .addAll(Collections.nCopies(valueCount, BOXED_NULLABLE))
+                        .build(),
+                customMethodHandle);
+    }
+
+    @UsedByGeneratedCode
+    public static Slice concatWsArray(Slice separator, Block block)
+    {
+        if (separator == null || block == null) {
+            return null;
+        }
+        Slice[] args = new Slice[block.getPositionCount()];
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            if (block.isNull(i)) {
+                args[i] = null;
+            }
+            else {
+                int sliceLength = block.getSliceLength(i);
+                args[i] = block.getSlice(i, 0, sliceLength);
+            }
+        }
+        return concatWs(separator, args);
+    }
+
+    @UsedByGeneratedCode
+    public static Slice concatWs(Slice separator, Slice[] values)
+    {
+        if (separator == null || values == null) {
```

Comment:
When would this happen? The separator argument is marked as non-nullable, and the values array is the result of collecting all the remaining arguments, so it should always be a non-null array, with possibly no elements.

Reply:
you're right, removed.